### PR TITLE
Add job table&actor table subscribe retry when gcs service restart

### PR DESCRIPTION
**Pull request by ffbin**
## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: src/ray/gcs/gcs_client/test/service_based_gcs_client_test.cc [hunk 1]
```diff
@@ -797,28 +812,66 @@ TEST_F(ServiceBasedGcsClientTest, TestErrorInfo) {
   ASSERT_TRUE(ReportJobError(error_table_data));
 }
 
-TEST_F(ServiceBasedGcsClientTest, TestDetectGcsAvailability) {
-  // Create job table data.
-  JobID add_job_id = JobID::FromInt(1);
-  auto job_table_data = Mocker::GenJobTableData(add_job_id);
+TEST_F(ServiceBasedGcsClientTest, TestJobTableReSubscribe) {
```
Comment (raulchen) [inline src/ray/gcs/gcs_client/test/service_based_gcs_client_test.cc:815]: Add a comment: `Test that subscription of the job table can still work when GCS server restarts.`

File: src/ray/rpc/gcs_server/gcs_rpc_client.h [hunk 2]
```diff
@@ -225,11 +228,20 @@ class GcsRpcClient {
     if (get_server_address_) {
       auto address = get_server_address_();
       Init(address.first, address.second, client_call_manager_);
+
+      if (Ping(address.first, address.second, 100)) {
```
Comment (raulchen) [inline src/ray/rpc/gcs_server/gcs_rpc_client.h:232]: What if this ping fails? 
I think we should either retry or crash. We shouldn't proceed without re-establishing the subscriptions.

File: src/ray/rpc/gcs_server/gcs_rpc_client.h [hunk 3]
```diff
@@ -225,11 +228,20 @@ class GcsRpcClient {
...
+        if (reconnected_callback_) {
+          reconnected_callback_();
+        }
+      }
     }
   }
 
   ClientCallManager &client_call_manager_;
   std::function<std::pair<std::string, int>()> get_server_address_;
+  std::function<void()> reconnected_callback_;
```
Comment (raulchen) [inline src/ray/rpc/gcs_server/gcs_rpc_client.h:244]: document

File: src/ray/gcs/gcs_client/service_based_accessor.cc [hunk 4]
```diff
@@ -352,6 +371,17 @@ Status ServiceBasedActorInfoAccessor::AsyncGetCheckpointID(
   return Status::OK();
 }
 
+Status ServiceBasedActorInfoAccessor::AsyncReSubscribe() {
+  RAY_LOG(INFO) << "Actor info resubscribe once.";
```
Comment (raulchen) [inline src/ray/gcs/gcs_client/service_based_accessor.cc:375]:
```suggestion
  RAY_LOG(INFO) << "Reestablishing subscription for actor info.";
```

File: src/ray/gcs/gcs_client/service_based_accessor.h [hunk 5]
```diff
@@ -43,7 +45,11 @@ class ServiceBasedJobInfoAccessor : public JobInfoAccessor {
       const SubscribeCallback<JobID, JobTableData> &subscribe,
       const StatusCallback &done) override;
 
+  Status AsyncReSubscribe() override;
+
  private:
+  SubscribeOperation subscribe_operation_;
```
Comment (raulchen) [inline src/ray/gcs/gcs_client/service_based_accessor.h:51]: `// Save the subscribe operation in this function, so we can call it again when GCS restarts from a failure.`

File: src/ray/gcs/accessor.h [hunk 6]
```diff
@@ -151,6 +151,8 @@ class ActorInfoAccessor {
       const ActorID &actor_id,
       const OptionalItemCallback<rpc::ActorCheckpointIdData> &callback) = 0;
 
+  virtual Status AsyncReSubscribe() = 0;
```
[the remark this answers is not in the document]
Comment (ffbin) [reply src/ray/gcs/accessor.h:154]: fixed

File: src/ray/gcs/gcs_client/test/service_based_gcs_client_test.cc [hunk 7]
```diff
@@ -63,6 +63,21 @@ class ServiceBasedGcsClientTest : public RedisServiceManagerForTest {
     FlushAll();
   }
 
+  void RestartGcsServer() {
+    RAY_LOG(INFO) << "Initializing GCS service, port = " << gcs_server_->GetPort();
```
[the remark this answers is not in the document]
Comment (ffbin) [reply src/ray/gcs/gcs_client/test/service_based_gcs_client_test.cc]: fixed

File: src/ray/gcs/gcs_client/test/service_based_gcs_client_test.cc (hunk 1 above)
Comment (ffbin) [reply src/ray/gcs/gcs_client/test/service_based_gcs_client_test.cc:815]: fixed

File: src/ray/rpc/gcs_server/gcs_rpc_client.h (hunk 2 above)
Comment (ffbin) [reply src/ray/rpc/gcs_server/gcs_rpc_client.h:235]: if this ping fails, we will retry.

File: src/ray/rpc/gcs_server/gcs_rpc_client.h (hunk 3 above)
Comment (ffbin) [reply src/ray/rpc/gcs_server/gcs_rpc_client.h:252]: fixed

File: src/ray/gcs/gcs_client/service_based_accessor.cc (hunk 4 above)
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_accessor.cc]: fixed

File: src/ray/gcs/gcs_client/service_based_accessor.h (hunk 5 above)
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_accessor.h:55]: fixed

File: src/ray/rpc/gcs_server/gcs_rpc_client.h [hunk 8]
```diff
@@ -228,11 +231,25 @@ class GcsRpcClient {
     if (get_server_address_) {
       auto address = get_server_address_();
       Init(address.first, address.second, client_call_manager_);
+
+      if (Ping(address.first, address.second, 100)) {
+        RAY_LOG(INFO) << "Target rpc server " << address.first << ":" << address.second
```
[the remark this answers is not in the document]
Comment (ffbin) [reply src/ray/rpc/gcs_server/gcs_rpc_client.h]: This log will be printed only when reconnection is successful,  so info log is easy to locate problems.

File: src/ray/rpc/gcs_server/gcs_rpc_client.h [hunk 9]
```diff
@@ -225,14 +229,42 @@ class GcsRpcClient {
...
     if (get_server_address_) {
-      auto address = get_server_address_();
-      Init(address.first, address.second, client_call_manager_);
+      std::pair<std::string, int> address;
+      int index = 0;
+      for (; index < RayConfig::instance().ping_gcs_rpc_server_max_retries(); ++index) {
+        address = get_server_address_();
+        if (Ping(address.first, address.second, 100)) {
+          RAY_LOG(INFO) << "Target rpc server " << address.first << ":" << address.second
+                        << " is valid.";
```
Comment (raulchen) [inline src/ray/rpc/gcs_server/gcs_rpc_client.h:240]: "Reconnected to GCS server: xxx." would be more clear.

File: src/ray/rpc/gcs_server/gcs_rpc_client.h [hunk 10]
```diff
@@ -225,14 +229,42 @@ class GcsRpcClient {
...
+
+      if (index < RayConfig::instance().ping_gcs_rpc_server_max_retries()) {
+        Init(address.first, address.second, client_call_manager_);
+        if (reconnected_callback_) {
+          reconnected_callback_();
+        }
+      } else {
+        RAY_LOG(FATAL) << "Target rpc server " << address.first << ":" << address.second
+                       << " is invalid.";
+      }
```
Comment (raulchen) [inline src/ray/rpc/gcs_server/gcs_rpc_client.h:254]: "Couldn't reconnect to GCS server. The last attempted GCS server address was xxx."

File: src/ray/rpc/gcs_server/gcs_rpc_client.h [hunk 11]
```diff
@@ -225,14 +229,42 @@ class GcsRpcClient {
...
 
   void Reconnect() {
+    absl::MutexLock lock(&mutex_);
     if (get_server_address_) {
-      auto address = get_server_address_();
-      Init(address.first, address.second, client_call_manager_);
+      std::pair<std::string, int> address;
+      int index = 0;
+      for (; index < RayConfig::instance().ping_gcs_rpc_server_max_retries(); ++index) {
+        address = get_server_address_();
```
Comment (raulchen) [inline src/ray/rpc/gcs_server/gcs_rpc_client.h:237]: print a debug log here: "Attempt to reconnect to GCS server: xxx".

File: src/ray/rpc/gcs_server/gcs_rpc_client.h [hunk 12]
```diff
@@ -225,14 +229,42 @@ class GcsRpcClient {
...
 
+  absl::Mutex mutex_;
+
   ClientCallManager &client_call_manager_;
   std::function<std::pair<std::string, int>()> get_server_address_;
+  /// This function is triggered when RPC request is called.
+  /// We will ping the address of the GCS server, and if it is valid, this function will
+  /// be called. Currently, we will subscribe to GCS again in this function. If the
+  /// function call succeeds but the connection of GCS server fails, it will try again and
+  /// subscription function needs to be idempotent.
```
Comment (raulchen) [inline src/ray/rpc/gcs_server/gcs_rpc_client.h:266]: /// The callback that will be called when we reconnect to GCS server.
/// Currently, we use this function to reestablish subscription to GCS. 
/// Note, we use `ping` to detect whether the reconnection is successful. If the ping succeeds but the
/// RPC connection fails, this function might be called called again. So it needs to be idempotent.

**Review by raulchen — APPROVED** [review]
LGTM. Only some minor comments.

File: src/ray/rpc/gcs_server/gcs_rpc_client.h (hunk 12 above)
Comment (ffbin) [reply src/ray/rpc/gcs_server/gcs_rpc_client.h]: fixed

File: src/ray/rpc/gcs_server/gcs_rpc_client.h (hunk 11 above)
Comment (ffbin) [reply src/ray/rpc/gcs_server/gcs_rpc_client.h:237]: fixed

File: src/ray/rpc/gcs_server/gcs_rpc_client.h (hunk 10 above)
Comment (ffbin) [reply src/ray/rpc/gcs_server/gcs_rpc_client.h]: fixed

File: src/ray/rpc/gcs_server/gcs_rpc_client.h (hunk 9 above)
Comment (ffbin) [reply src/ray/rpc/gcs_server/gcs_rpc_client.h]: fixed

**Pull request merged**